go/libkb/kbpackets.go

```diff
@@ -16,10 +13,147 @@ import (
 	"fmt"
 	"io"
 
-	keybase1 "github.com/keybase/client/go/protocol/keybase1"
 	"github.com/keybase/go-codec/codec"
 )
 
+type Packetable interface {
+	GetTagAndVersion() (PacketTag, PacketVersion)
+}
+
+func EncodePacket(p Packetable, w io.Writer) error {
+	packet, err := newKeybasePacket(p)
+	if err != nil {
+		return err
+	}
+	return packet.encodeTo(w)
+}
+
+func EncodePacketToBytes(p Packetable) ([]byte, error) {
+	packet, err := newKeybasePacket(p)
+	if err != nil {
+		return nil, err
+	}
+	return packet.encode()
+}
+
+func EncodePacketToArmoredString(p Packetable) (string, error) {
+	packet, err := newKeybasePacket(p)
+	if err != nil {
+		return "", err
+	}
+	return packet.armoredEncode()
+}
+
+func DecodePacket(decoder *codec.Decoder, body Packetable) error {
+	// TODO: Do something with the version too?
+	tag, _ := body.GetTagAndVersion()
+	p := KeybasePacket{
+		Body: body,
+	}
+	err := decoder.Decode(&p)
+	if err != nil {
+		return err
+	}
+
+	if p.Tag != tag {
+		return UnmarshalError{ExpectedTag: p.Tag, Tag: tag}
+	}
+
+	// TODO: Figure out a way to do the same reencode check as in
+	// DecodePacketFromBytes.
+
+	return p.checkHash()
+}
+
+func DecodePacketFromBytes(data []byte, body Packetable) error {
+	ch := codecHandle()
+	decoder := codec.NewDecoderBytes(data, ch)
+
+	// TODO: Do something with the version too?
+	tag, _ := body.GetTagAndVersion()
+	p := KeybasePacket{
```

Comment:
we can just decode directly into the Packetable object, saving the reencode we used to do. (This generalizes the optimization that was done for NaclSigInfo).